packages/jsii-pacmak/lib/targets/java.ts

```diff
@@ -9,13 +9,132 @@ import { Generator } from '../generator';
 import logging = require('../logging');
 import { md2html } from '../markdown';
 import { PackageInfo, Target } from '../target';
-import { shell } from '../util';
+import { shell, Scratch } from '../util';
 import { VERSION, VERSION_DESC } from '../version';
+import { TargetBuilder, BuildOptions, allOutputDirectoriesTheSame, OneByOneBuilder } from '../builder';
+import { JsiiModule } from '../packaging';
 
 /* eslint-disable @typescript-eslint/no-var-requires */
 const spdxLicenseList = require('spdx-license-list');
 /* eslint-enable @typescript-eslint/no-var-requires */
 
+/**
+ * Build Java packages in parallel, by generating an aggregate POM
+ */
+export class JavaBuilder implements TargetBuilder {
+  private readonly targetName = 'java';
+
+  public async buildModules(modules: JsiiModule[], options: BuildOptions): Promise<void> {
+    if (modules.length === 0) { return; }
+
+    // We can only do the optimized build if '-o' was specified, which we will notice
+    // as all module outputdirectories being the same. (Maybe we can build to per-package
+    // dist dirs as well, but this is the smallest delta from what we had and we will
+    // always specify the output dir anyway).
+    if (!allOutputDirectoriesTheSame(modules)) {
+      logging.warn('Single output directory not specified, doing (slower) one-by-one build for Java');
+      await new OneByOneBuilder(this.targetName, Java).buildModules(modules, options);
+      return;
+    }
+
+    const singleOutputDir = this.finalOutputDir(modules[0], options);
+
+    const moduleDirectories = [];
+
+    for (const module of modules) {
+      moduleDirectories.push(await this.generateModuleCode(module, options, options.codeOnly));
+    }
+
+    if (!options.codeOnly && modules.length > 0) {
+      // Need a module to get a target
+      const pomDirectory = await this.generateAggregatePom(moduleDirectories);
+      const target = this.makeTarget(modules[0], options);
+
+      await target.build(pomDirectory, singleOutputDir);
+    }
+  }
+
+  private async generateModuleCode(module: JsiiModule, options: BuildOptions, finalDirectory?: boolean): Promise<Scratch<void>> {
+    const target = this.makeTarget(module, options);
+
+    const srcDir = finalDirectory
+      ? Scratch.fake(this.finalOutputDir(module, options), undefined)
+      : await Scratch.make(_ => undefined);
+
+    logging.debug(`Generating ${this.targetName} code into ${srcDir.directory}`);
+    await target.generateCode(srcDir.directory, module.tarball);
+
+    return srcDir;
+  }
+
+  private async generateAggregatePom(sourceDirectories: Array<Scratch<void>>) {
+    const parentDir = this.findSharedParentDirectory(sourceDirectories.map(s => s.directory));
+
+    const aggregatePom = xmlbuilder.create({
```

Comment:
Would be interesting if this mode could also produce a Maven `BOM` file!

Reply:
What's that?